designs/UserWideConfiguration-AdditionalExtensibilityFolders.md

```diff
@@ -0,0 +1,78 @@
+
+# Additional user wide configuration options
+
+* Status: In Review
+* Author(s): [Nikolche Kolev](https://github.com/nkolev92)
+* Issue: [9394](https://github.com/NuGet/Home/issues/9394) NuGet should support multiple config files in %APPDATA%\NuGet directory
+
+## Problem Background
+
+The XAML and WinForms teams want to allow control library installers to add user-specific NuGet fallback folders by having their SDK installers drop config files in %APPDATA%\NuGet. On uninstall this file would be deleted.
+
+We considered two alternatives which I mentioned as workarounds above:
+
+1. Require SDK installers to drop config files in %ProgramFiles(x86)%\NuGet\Config instead. However, we've gotten feedback from control vendors that they would like the option of installing without elevation which prevents the installer from being able to write to this directory.
+1. Require SDK installers to modify the one %APPDATA%\NuGet\NuGet.config file. However this seems error prone.

```

Comment:
Why is this error prone? Adding one line and then removing it upon uninstall should be easy IMO. Similar to adding and removing sources/feeds.

Reply:
My understanding is they're going to be adding a fallback folder, not package source. So if the user's config file doesn't already contain a fallback folders element, then it's 3 XML elements that needs to be added, and then there's the question whether the fallbackFolders element should be deleted if the last fallback folder is removed from that list.

If the fallbackFolders list contains multiple entries and the uninstaller has a bug, it might delete the wrong line in the file, or might delete all the fallback folders, not just its own. There's less risk that the uninstaller deletes the wrong file, or overwrites something it shouldn't on install.

Depending on what APIs the control vendor uses in their setup program, it might strip out XML comments from the file, or change whitespace. While this won't cause problems, it might not be what the customer wants, particularly for a file that is not version controlled.

Installer technologies are optimised for writing files and uninstall to delete files, so that's the easiest thing to do. Adding logic to edit existing files, while possible, increases the complexity for control vendors to use this feature.

It would be better for performance if we didn't load `*.config` from anywhere, but the risks and additional complexity for anyone who wants to use this feature doesn't feel worthwhile to me. We can minimise perf impact by loading in parallel if we're not already.

Reply:
How does a Fallback folder line look in the nuget.config? If we can provide an API to add and remove the fallback folder, given a unique name, should it still be a problem. My thinking is to abstain from adding more moving parts to the nuget config management.

Reply:
```xml
<fallbackPackageFolders>
   <add key="XYZ Offline Packages" value="C:\somePath\someFolder\"/>
</fallbackPackageFolders>
```

from https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file#fallbackpackagefolders-section. 

Given that the installation of these configs is a 3rd party experience, they are free to use our libraries. We have them, but we don't know if the installer is written in .NET. 

Regarding the error prone part, editing onfigs has caused unexpected problems for customers in the past. 
There are some occasions in which NuGet itself trying to edit the config caused issues. 

@rrelyea might know more about the history of those issues.

Reply:
It looks almost identical to the packageSources section: https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file#fallbackpackagefolders-section.

Even if we create an API (presumably in .NET, although we technically already have APIs, in NuGet.Configuration.dll), I have no idea how installers would call the API. Can MSI's natively execute .NET assemblies, or would the MSI author have to jump through additional hoops to copy .NET assemblies to a temporary location, run, then clean up this temporary stuff after itself? What about installers that are not MSIs?